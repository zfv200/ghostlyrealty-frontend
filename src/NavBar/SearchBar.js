import React from 'react'
import { connect } from 'react-redux'
import { searchSite } from '../actions/actions'
import { withRouter } from 'react-router-dom';
import { addBlankSearchError, clearBlankSearchError } from '../actions/actions'
import BlankSearch from './BlankSearch'
import styles from './NavBar.css.js'

// TODO: is there a good way to abstract the blank search functionality to wrap all the search forms?


export class SearchBar extends React.Component {
  constructor(){
    super()

    this.state={
      searchTerm: ''
    }
  }

  handleChange = (e) => {
    this.setState({
      searchTerm: e.target.value
    })
  }

  handleSubmit = (e) => {
    e.preventDefault()
    if (this.state.searchTerm===''){
      return this.props.addBlankSearchError()
    } else {
      this.props.clearBlankSearchError()
    }
    this.props.searchSite(this.state.searchTerm)
    this.setState({
      searchTerm: ''
    })
    this.props.history.push('/results')
  }

  renderBlankSearchError = () => {
    return this.props.blankSearchError ? <BlankSearch /> : null
  }

  render(){
    return (
      <div>
        <div>
          {this.renderBlankSearchError()}
          <form id="nav-search-bar-form" onSubmit={this.handleSubmit}>
            <input id="nav-search-bar" style={styles.siteSearchBar} placeholder="house or agent name" value={this.state.searchTerm} type="text" onChange={this.handleChange}/>
          </form>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    blankSearchError: state.searchReducer.blankSearch
  }
}

export default connect(mapStateToProps, {searchSite, addBlankSearchError, clearBlankSearchError})(withRouter(SearchBar))
