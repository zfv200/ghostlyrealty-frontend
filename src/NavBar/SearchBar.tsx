import React from 'react'
import { connect } from 'react-redux'
import { searchSite } from '../actions/actions'
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { addBlankSearchError, clearBlankSearchError } from '../actions/actions'
import BlankSearch from './BlankSearch'
import styles from './NavBar.css.js'

// TODO: is there a good way to abstract the blank search functionality to wrap all the search forms?
// import { rootReducer } from '../index.js'

interface StateProps {
  blankSearchError: boolean
}

interface DispatchProps {
  searchSite: (searchTerm: string) => void,
  addBlankSearchError: () => void,
  clearBlankSearchError: () => void
}

type Props = StateProps & DispatchProps & RouteComponentProps<any>

interface State {
  searchTerm: string
}

export class SearchBar extends React.Component<Props, State> {

  state={
    searchTerm: ''
  }

  handleChange = (e: React.FormEvent<HTMLInputElement>): void => {
    const target = e.target as HTMLTextAreaElement;
    this.setState({
      searchTerm: target.value
    })
  }

  handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    if (this.state.searchTerm===''){
      this.props.addBlankSearchError()
    } else {
      this.props.clearBlankSearchError()
    }
    this.props.searchSite(this.state.searchTerm)
    this.setState({
      searchTerm: ''
    })
    this.props.history.push('/results')
  }

  renderBlankSearchError = () => {
    return this.props.blankSearchError ? <BlankSearch /> : null
  }

  render(){
    return (
      <div>
        {this.renderBlankSearchError()}
        <form id="nav-search-bar-form" onSubmit={(e)=>this.handleSubmit(e)}>
          <input id="nav-search-bar" style={styles.siteSearchBar} placeholder="house or agent name" value={this.state.searchTerm} type="text" onChange={(e)=>this.handleChange(e)}/>
        </form>
      </div>
    )
  }
}

const mapStateToProps = (state: any): StateProps => {
  return {
    blankSearchError: state.searchReducer.blankSearch
  }
}

export default connect(mapStateToProps, {searchSite, addBlankSearchError, clearBlankSearchError})(withRouter(SearchBar))
