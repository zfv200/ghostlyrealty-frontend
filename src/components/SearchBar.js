import React from 'react'
import { connect } from 'react-redux'
import { searchSite } from '../actions/actions'
import { withRouter } from 'react-router-dom';

class SearchBar extends React.Component {
  constructor(){
    super()

    this.state={
      searchTerm: ''
    }
  }

  handleChange = (e) => {
    this.setState({
      searchTerm: e.target.value
    })
  }

  handleSubmit = (e) => {
    e.preventDefault()
    this.props.searchSite(this.state.searchTerm)
    this.props.history.push('/results')
  }

  render(){
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <input type="text" onChange={this.handleChange}/>
        </form>
      </div>
    )
  }
}

export default connect(null, {searchSite})(withRouter(SearchBar))
