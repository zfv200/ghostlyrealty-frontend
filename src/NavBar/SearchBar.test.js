import React from 'react'
import { shallow } from 'enzyme'
import { SearchBar } from './SearchBar'

let mockProps = {
  history: [],
  addBlankSearchError: jest.fn(),
  clearBlankSearchError: jest.fn(),
  searchSite: jest.fn()
}

describe('searchBar', ()=>{
  let searchBar = shallow(<SearchBar {...mockProps}/>)
  const fakeEvent = { preventDefault: () => console.log('preventDefault') };

  it('renders correctly without crashing', ()=>{

    expect(searchBar).toMatchSnapshot()
  })

  it('updates the input when query is typed', ()=>{
    let searchBar = shallow(<SearchBar />)

    searchBar.find('[id="nav-search-bar"]').simulate('change', { target: { value: 'Hello' } })
    expect(searchBar.state()).toEqual({
      searchTerm: 'Hello'
    })
  })

  it('calls the blank search error when submitted with empty input', ()=>{
    searchBar.find('[id="nav-search-bar-form"]').simulate('submit', fakeEvent);

    expect(mockProps.addBlankSearchError).toBeCalled()
  })

  it('pushes the correct path to the history object when submitted', ()=>{
    searchBar.find('[id="nav-search-bar"]').simulate('change', { target: { value: 'Hello' } })
    searchBar.find('[id="nav-search-bar-form"]').simulate('submit', fakeEvent);
    expect(mockProps.history).toEqual(['/results'])
  })

  it('executes siteSearch prop function when submitted', ()=>{
    searchBar.find('[id="nav-search-bar"]').simulate('change', { target: { value: 'Hello' } })
    searchBar.find('[id="nav-search-bar-form"]').simulate('submit', fakeEvent);
    expect(mockProps.searchSite).toBeCalled()
  })

  it('clears out the blank search error when query is provided after error', ()=>{
    //left off here
    searchBar.find('[id="nav-search-bar-form"]').simulate('submit', fakeEvent);
    expect(mockProps.addBlankSearchError).toBeCalled()
    searchBar.find('[id="nav-search-bar"]').simulate('change', { target: { value: 'Hello' } })
    expect(mockProps.clearBlankSearchError).toBeCalled()

  })

  it('clears out the search value/state when site is successfully searched', ()=>{
    searchBar.find('[id="nav-search-bar"]').simulate('change', { target: { value: 'Hello' } })
    searchBar.find('[id="nav-search-bar-form"]').simulate('submit', fakeEvent);
    expect(searchBar.state()).toEqual({
      searchTerm: ''
    })
  })

})
