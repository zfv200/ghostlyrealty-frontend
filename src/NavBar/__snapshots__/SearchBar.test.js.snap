// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchBar renders correctly without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchBar
    addBlankSearchError={[MockFunction]}
    clearBlankSearchError={[MockFunction]}
    history={Array []}
    searchSite={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <form
          id="nav-search-bar-form"
          onSubmit={[Function]}
        >
          <input
            id="nav-search-bar"
            onChange={[Function]}
            placeholder="house or agent name"
            style={
              Object {
                "height": "40px",
                "marginRight": "120px",
                "width": "400px",
              }
            }
            type="text"
            value=""
          />
        </form>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          null,
          <form
            id="nav-search-bar-form"
            onSubmit={[Function]}
          >
            <input
              id="nav-search-bar"
              onChange={[Function]}
              placeholder="house or agent name"
              style={
                Object {
                  "height": "40px",
                  "marginRight": "120px",
                  "width": "400px",
                }
              }
              type="text"
              value=""
            />
          </form>,
        ],
      },
      "ref": null,
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <input
              id="nav-search-bar"
              onChange={[Function]}
              placeholder="house or agent name"
              style={
                Object {
                  "height": "40px",
                  "marginRight": "120px",
                  "width": "400px",
                }
              }
              type="text"
              value=""
            />,
            "id": "nav-search-bar-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "nav-search-bar",
              "onChange": [Function],
              "placeholder": "house or agent name",
              "style": Object {
                "height": "40px",
                "marginRight": "120px",
                "width": "400px",
              },
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "type": "form",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <form
            id="nav-search-bar-form"
            onSubmit={[Function]}
          >
            <input
              id="nav-search-bar"
              onChange={[Function]}
              placeholder="house or agent name"
              style={
                Object {
                  "height": "40px",
                  "marginRight": "120px",
                  "width": "400px",
                }
              }
              type="text"
              value=""
            />
          </form>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            null,
            <form
              id="nav-search-bar-form"
              onSubmit={[Function]}
            >
              <input
                id="nav-search-bar"
                onChange={[Function]}
                placeholder="house or agent name"
                style={
                  Object {
                    "height": "40px",
                    "marginRight": "120px",
                    "width": "400px",
                  }
                }
                type="text"
                value=""
              />
            </form>,
          ],
        },
        "ref": null,
        "rendered": Array [
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                id="nav-search-bar"
                onChange={[Function]}
                placeholder="house or agent name"
                style={
                  Object {
                    "height": "40px",
                    "marginRight": "120px",
                    "width": "400px",
                  }
                }
                type="text"
                value=""
              />,
              "id": "nav-search-bar-form",
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "nav-search-bar",
                "onChange": [Function],
                "placeholder": "house or agent name",
                "style": Object {
                  "height": "40px",
                  "marginRight": "120px",
                  "width": "400px",
                },
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "form",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
